builtins/web/form-data/form-data-encoder.cpp

```diff
@@ -0,0 +1,714 @@
+  auto impl = as_impl(self);
+
+  // Try to fill the buffer
+  while (total < bufsz && !finished) {
+    auto subspan = buf.subspan(total);
+    auto stream = StreamContext(entries, subspan);
+
+    if (!impl->read_next(cx, stream)) {
+      return false;
+    }
+
+    total += stream.read;
+    finished = stream.done;
+  }
+
+  // Delay reporting done to produce a separate empty chunk.
+  *done = finished && total == 0;
+  *read = total;
+  return true;
+}
+
+std::string MultipartFormData::boundary(JSObject *self) {
+  MOZ_ASSERT(is_instance(self));
+  auto impl = as_impl(self);
+  MOZ_ASSERT(impl);
+
+  return impl->boundary();
+}
+
+MultipartFormDataImpl *MultipartFormData::as_impl(JSObject *self) {
+  MOZ_ASSERT(is_instance(self));
+  return reinterpret_cast<MultipartFormDataImpl *>(
+      JS::GetReservedSlot(self, Slots::Inner).toPrivate());
+}
+
+JSObject *MultipartFormData::form_data(JSObject *self) {
+  MOZ_ASSERT(is_instance(self));
+  return &JS::GetReservedSlot(self, Slots::Form).toObject();
+}
+
+std::optional<size_t> MultipartFormData::query_length(JSContext *cx, HandleObject self) {
+  RootedObject obj(cx, form_data(self));
+
+  auto entries = FormData::entry_list(obj);
+  auto impl = as_impl(self);
+
+  return impl->query_length(cx, entries);
+}
+
+JSObject *MultipartFormData::encode_stream(JSContext *cx, HandleObject self) {
+  RootedObject reader(cx, BufReader::create(cx, self, read));
+  if (!reader) {
+    return nullptr;
+  }
+
+  RootedObject native_stream(cx, BufReader::stream(reader));
+  RootedObject default_stream(cx, NativeStreamSource::stream(native_stream));
+
+  return default_stream;
+}
+
+JSObject *MultipartFormData::create(JSContext *cx, HandleObject form_data) {
+  JS::RootedObject self(cx, JS_NewObjectWithGivenProto(cx, &class_, proto_obj));
+  if (!self) {
+    return nullptr;
+  }
+
+  if (!FormData::is_instance(form_data)) {
+    return nullptr;
+  }
+
+  auto res = host_api::Random::get_bytes(12);
+  if (auto *err = res.to_err()) {
+    return nullptr;
+  }
+
+  // The requirements for boundary are (https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.1):
+  // Boundary delimiters must not appear within the encapsulated material, and must be no longer than
+  // 70 characters, not counting the two leading hyphens and consist of bcharsnospace characters,
+  // where EBNF for bcharsnospace is as follows:
+  //
+  // bcharsnospace := DIGIT / ALPHA / "'" / "(" / ")" / "+" / "_" / "," / "-" / "." / "/" / ":" / "=" / "?"
+  //
+  // e.g.:
+  // This implementation: --BoundaryjXo5N4HEAXWcKrw7
+  // WebKit: ----WebKitFormBoundaryhpShnP1JqrBTVTnC
+  // Gecko:  ----geckoformboundary8c79e61efa53dc5d441481912ad86113
+  auto bytes = std::move(res.unwrap());
+  auto bytes_str = std::string_view((char *)(bytes.ptr.get()), bytes.size());
+  auto base64_str = base64::forgivingBase64Encode(bytes_str, base64::base64EncodeTable);
+
+  auto boundary = fmt::format("--Boundary{}", base64_str);
```

Comment:
Maybe `StarlingMonkeyFormBoundary` prefix would be better?

Reply:
Not sure it makes too much of a difference, but I guess it wouldn't hurt to match other implementations' approach. Up to you though!